webapp/src/cloud.ts

```diff
@@ -0,0 +1,196 @@
+import * as core from "./core";
+import * as auth from "./auth";
+import * as ws from "./workspace";
+import * as data from "./data";
+import * as workspace from "./workspace";
+
+type Version = pxt.workspace.Version;
+type File = pxt.workspace.File;
+type Header = pxt.workspace.Header;
+type ScriptText = pxt.workspace.ScriptText;
+type WorkspaceProvider = pxt.workspace.WorkspaceProvider;
+
+import U = pxt.Util;
+
+const state = {
+    uploadCount: 0,
+    downloadCount: 0
+};
+
+type CloudProject = {
+    id: string;
+    header: string;
+    text: string;
+    version: string;
+};
+
+async function listAsync(): Promise<Header[]> {
+    return new Promise(async (resolve, reject) => {
+        const result = await auth.apiAsync<CloudProject[]>("/api/user/project");
+        if (result.success) {
+            const userId = auth.user()?.id;
+            const headers = result.resp.map(proj => {
+                const header = JSON.parse(proj.header);
+                header.cloudUserId = userId;
+                header.cloudVersion = proj.version;
+                header.cloudCurrent = true;
+                return header;
+            });
+            resolve(headers);
+        } else {
+            reject(new Error(result.errmsg));
+        }
+    });
+}
+
+function getAsync(h: Header): Promise<File> {
+    return new Promise(async (resolve, reject) => {
+        const result = await auth.apiAsync<CloudProject>(`/api/user/project/${h.id}`);
+        if (result.success) {
+            const userId = auth.user()?.id;
+            const project = result.resp;
+            const header = JSON.parse(project.header);
+            const text = JSON.parse(project.text);
+            const version = project.version;
+            const file: File = {
+                header,
+                text,
+                version
+            };
+            file.header.cloudCurrent = true;
+            file.header.cloudVersion = file.version;
+            file.header.cloudUserId = userId;
+            resolve(file);
+        } else {
+            reject(new Error(result.errmsg));
+        }
+    });
+}
+
+function setAsync(h: Header, prevVersion: Version, text?: ScriptText): Promise<Version> {
```

Comment:
I think we will need to serialize save operations so that we always send the latest `prevVersion` (it maps to the cosmos record's etag). We want to avoid having multiple saves in flight or we'll get etag rejections.

Related: etag check needs to be enabled on the backend. Code is there but commented out at the moment.

Reply:
I think we could throttle these on the client. i.e. don't send every individual save but coalesce them and sync periodically. The client saves the project after every change so this could be important to keep from flooding the backend.